aiida/engine/processes/calcjobs/calcjob.py

```diff
@@ -593,12 +593,9 @@ def presubmit(self, folder: Folder) -> CalcInfo:
                 raise PluginInternalError('CalcInfo should have the information of the code to be launched')
             this_code = load_node(code_info.code_uuid, sub_classes=(Code,))
 
-            this_withmpi = code_info.withmpi  # to decide better how to set the default
-            if this_withmpi is None:
-                if len(calc_info.codes_info) > 1:
-                    raise PluginInternalError('For more than one code, it is necessary to set withmpi in codes_info')
-                else:
-                    this_withmpi = self.node.get_option('withmpi')
+            this_withmpi = self.node.get_option('withmpi')  # set code mpi, default set to as calcjob one
+            if code_info.withmpi:
```

Comment:
If I understand correctly, there are two ways to set `withmpi`: on the process itself through the `metadata.options.withmpi` input and through the `CodeInfo` that is created by the plugin itself, where the latter should be leading if both are defined.

If that analysis is correct, than I think this code is not correct. By checking just `if code_info.withmpi` the actual value will not be overridden even if it is set to `False`. However, this seems like a good use-case from a plugin perspective. If the plugin needs to run two codes, for which one _always_ needs to be run without mpi, i.e. `code_info.withmpi == False`, then that should override the value taken from `self.node.get_option('withmpi')`. That is to say, `False` is also an actual value that should not be ignored. So should the conditional not be `if code_info.withmpi is not None`?

To make this code a lot clearer, I would propose writing it as follows, with comment:
```suggestion

            # To determine whether this code should be run with MPI enabled, we get the value that was set in the inputs
            # of the entire process, which can then be overwritten by the value from the `CodeInfo`. This allows plugins
            # to force certain codes to run without MPI, even if the user wants to run all codes with MPI whenever
            # possible. This use case is typically useful for `CalcJob`s that consist of multiple codes where one or
            # multiple codes always have to be executed without MPI.

            withmpi = self.node.get_option('withmpi')

            # Override the value of `withmpi` with that of the `CodeInfo` if and only if it is set
            if code_info.withmpi is not None:
                withmpi = code_info.withmpi
```

Reply:
Uhm, I totally missed this on my pass, my bad 😕.

Follow up question: we want `code_info.withmpi` to **always** override the calcjob setting? Or if the user has explicitly set `withmpi` to be `False` then all codes executed should actually respect that? (in which case setting `code_info.withmpi = True` or leaving `code_info.withmpi = None` would have the same effect) Or the idea is that if a plugin sets `code_info.withmpi = True` it means that the code can only be ran with mpi?

Maybe we can make a table of expected results and intent and check these in a test?

|  | [Calcjob] True | [Calcjob] False |
| :---  | :---:  | :---:  |
| [Code] True (the plugin/code can only be run with mpi?) | Run with MPI | Incompatible? Raise? |
| [Code] False (this code does not use parallelism at all)| Run without MPI | Run without MPI |
| [Code] None (this code can run with or without MPI) | Run with MPI | Run without MPI |

Reply:
> Or if the user has explicitly set withmpi to be False then all codes executed should actually respect that?

In the example that I sketched, this would be problematic. Imagine a `CalcJob` that executes two codes: a small pre-processing script that _has_ to be executed serially or it will crash, and then a normal executable (like `pw.x`) that can be executed serially or with MPI depending on how it is compiled. In this case, the `metadata.options.withmpi` is the tool for the user to toggle the behavior for the latter, but it should _not_ override the setting for the pre-processing script because it will fail.

I am not sure if this solution covers all use cases or whether there are exceptions where the user definitely needs to be able to override. However, if the plugin explicitly sets `CodeInfo.withmpi` than there must be a good reason. If it can be run with either setting, maybe the plugin should not set one value either way. Or we should change `metadata.options.withmpi` to be able to take a list, but that may become even more problematic.

Reply:
@ramirezfranciscof @sphuber thanks for making this more clear. This is an important catch! It is the right code that I intended to implement. 

Only if plugin developer sets `CodeInfo.withmpi` explicitly the `metadata.option.withmpi` will not override it through running `CalcJob`. The ultimate behaviour will be the case shown in the table above. Indeed that does not include the scenario that the plugin user what to override what was set in CodeInfo. I have no strong opinion on whether we should take this into account. 
But I want to mention that maybe we can consider this problem together with [setting default calcjob when configuring code #4991](https://github.com/aiidateam/aiida-core/issues/4991). Plugin user can only set one default CalcJob plugin for the code and this is the code that is implicitly called with `self.inputs.code` in `CalcJob`.  I would assume the plugin user only what to control the MPI behaviour of this particular code. And the other codes need to be explicitly passed as an InputPort in defining the CalcJob, which give plugin user an option to use the correct compiled code(MPI one or serial one) if the `withmpi` of this codeinfo is fixed by the plugin developer.  
Therefore, I prefer to have a detailed comment as shown by @sphuber and add information in documentation to warn the plugin developer to carefully set this `codeinfo.withmpi` in their plugin when there is more than one code.

Reply:
> In the example that I sketched, this would be problematic. Imagine a `CalcJob` that executes two codes: a small pre-processing script that _has_ to be executed serially or it will crash, and then a normal executable (like `pw.x`) that can be executed serially or with MPI depending on how it is compiled. In this case, the `metadata.options.withmpi` is the tool for the user to toggle the behavior for the latter, but it should _not_ override the setting for the pre-processing script because it will fail.

Yes, so I think you are describing what should happen in the table I included when `[Calcjob] True` and `[Code] False`. Here it makes sense that the `CodeInfo.withmpi` overrides the `metadata.options.withmpi` because of the reasons you already explained: even if the user is indicating it has MPI available for the main code called by the calculation, other auxiliary calls may not benefit from or not even be able to run in parallel.

My question was more what should happen in the reciprocal case, when `[Calcjob] False` and `[Code] True`. Here the user is specifically saying to NOT use MPI at all, is it still justified to override this and anyways use MPI on the `CodeInfo.withmpi = True` code? Or does this suggest a more fundamental incompatibility and thus should raise?

Reply:
My apologies @ramirezfranciscof, I gave it a glimpse and take for granted with my own mind. 
I reconsider your advice to raise here and start to like it. Maybe it can be a raising warning rather than an exception? But I still assume the developer of the plugin should be careful and know what they do when setting `codeinfo.withmpi`. 
Do you have a strong opinion on raising an exception or maybe a warning?

Reply:
My preference is typically on the versatility on the end user and clarity when we are overriding his intentions. So here I would say:

- I would issue a warning if then we let `metadata.option.withmpi` override `codeinfo.withmpi` (we let the user know the plugin designer set this to only be run with MPI for some reason but we allow them to proceed as they see fit)
- In the other case, i.e. if after the warning `codeinfo.withmpi` would override `metadata.option.withmpi`, then I would say it is preferable to raise. This setting is not being used for anything else as far as I know, if the user is setting it, it is specifically for this and we are overriding it. They wouldn't be able to do what they want anyways, so just raise saying "you are not allowed to set this".

Now as to which of those two I prefer, I'm not sure. I'm still a bit puzzled as to in what situation the plugin designer would set something to ONLY be run with MPI, it is a bit bizarre to me. I'll let you decide what is best for this (or @sphuber if he has another opinion).

---

NOTE:  I am now re-reading the previous logic to check all the cases of the table. The previous behavior was that it would first check the individual `codeinfo.withmpi` and only if it was `None` *and* if it was a single code, then it would check `metadata.option.withmpi`. So if there were multiple codes being called, then the plugin designer had to set whether they were run with MPI or not, and the user had no say in this. So before there was actually no way for the plugin designer to leave how to run some of the codes to "the criteria" of the user.

This gets more bizarre the more we look into it...🤔

Reply:
All the bizarre things come from there is only one binary flag that can be set from `metadata.option.withmpi` by plugin user when actually a list of flags need for multiple codes case, which is hard to settle at this PR.

I would like to suggest that we stop here and leave all complicity to the documentation. For plugins that involve more than one code, if the user wants to get full control in MPI mode of how to run codes, they are suggested to split the calcjob into parts with each one running with one code only.